api/ModelHandler.py

```diff
@@ -0,0 +1,314 @@
+modelColumns = ['chrom', 'chromStart', 'chromEnd', 'annotation', 'height']
+
+
+def getModel(data):
+    data['hub'] = data['name'].split('/')[0]
+
+    data['track'] = data['name'].split('/')[-1]
+
+    problems = th.getProblems(data)
+
+    output = []
+
+    lock = locks.getLock(data['name'])
+
+    lock.acquire()
+
+    for problem in problems:
+        # data path, hub path, track path, problem path
+        modelsPath = '%s%s/%s/%s-%s-%s/' % (pl.dataPath, data['hub'], data['track'],
+                                            problem['ref'], problem['start'], problem['end'])
+
+        summaryPath = '%smodelSummary.txt' % modelsPath
+
+        # Converter required because pandas will lose trailing 0's
+        try:
+            summary = pd.read_csv(summaryPath, sep='\t', header=None, converters={6: lambda x: str(x)})
+        except FileNotFoundError:
+            continue
+
+        summary.columns = summaryColumns
+
+        minError = summary[summary['errors'] == summary['errors'].min()]
+
+        # Uses first penalty with min label error
+        # This will favor the model with the lowest penalty, given that summary is sorted
+        penalty = minError['penalty'].iloc[0]
+
+        minErrorModelPath = '%s%s_Model.bedGraph' % (modelsPath, penalty)
+
+        # TODO: If no good model, do LOPART
+
+        model = loadModelFile(minErrorModelPath, data)
+
+        output.extend(model)
+
+    lock.release()
+
+    return output
+
+
+def loadModelFile(path, data):
+    refseq = data['ref']
+    start = data['start']
+    end = data['end']
+
+    output = []
+
+    with open(path) as model:
+        modelLine = model.readline()
+
+        while not modelLine == '':
+            lineVals = modelLine.split()
+            lineStart = int(lineVals[1])
+            lineEnd = int(lineVals[2])
+            lineLabel = lineVals[3]
+
+            lineIfStartIn = (lineStart >= start) and (lineStart <= end)
+            lineIfEndIn = (lineEnd >= start) and (lineEnd <= end)
+            wrap = (lineStart < start) and (lineEnd > end)
+
+            if (lineIfStartIn or lineIfEndIn or wrap) and lineLabel == 'peak':
+                lineHeight = float(lineVals[4])
+                output.append({"ref": refseq, "start": lineStart,
+                               "end": lineEnd, "score": lineHeight})
+
+            modelLine = model.readline()
+
+    return output
+
+
+def updateModelLabels(data, generate=True):
+    data['hub'] = data['name'].split('/')[0]
+
+    data['track'] = data['name'].split('/')[-1]
+
+    # This is the problems that the label update is in
+    problems = th.getProblems(data)
+
+    threads = []
+
+    for problem in problems:
+        # data path, hub path, track path, problem path
+        modelsPath = '%s%s/%s/%s-%s-%s/' % (pl.dataPath, data['hub'], data['track'],
+                                            problem['ref'], problem['start'], problem['end'])
```

Comment:
this is identical to code above.. need to refactor to avoid repetition, even better use a database instead of file system for storage. Did you read/understand my BerkeleyDB database classes in https://github.com/tdhock/SegAnnDB/blob/master/plotter/db.py ?? why not use that?

Reply:
Non factor with DB Implementation